File: packages/core/ts/src/engine.ts
// ConversationEngine — the heart of botiva. Server-side symmetric of the
// chativa ChatEngine, playing the role DirectLine's connector service plays
// for BotFramework:
//
//   client(s) ⇄ transport connector ⇄ ConversationEngine ⇄ Runtime (LangGraph/...)
//                                          │
//                        ExtensionRegistry │ StateStore │ HistoryStore
//
// Responsibilities:
//   • Identity: userId (stable across devices) / conversationId (resumable) /
//     connectionId (one socket). Any number of connections may attach to the
//     same conversation — every outbound frame is fanned out to all of them.
//   • Resume: persistent frames get a monotonic seq; a client reconnecting
//     with a watermark receives everything it missed (DirectLine-style).
//   • Turns: one turn at a time per conversation; runtime events are merged
//     with botivaEmit() events, run through the extension chain, mapped to
//     wire frames, persisted and broadcast.
//   • HITL: interrupt events are stored as the conversation's pending
//     interrupt; the user's next message resumes the runtime with it.
//
// Scaling note: the engine itself is stateless apart from the live-connection
// registry and the per-conversation turn lock, both process-local. Multi
// instance deployments need sticky sessions (or a store-based lock) plus
// Redis-backed StateStore/HistoryStore.

import { randomUUID } from "node:crypto";
import { AsyncQueue } from "./queue.js";
import { ExtensionRegistry, type Extension } from "./extensions.js";
import {
    ConversationStore,
    MemoryStateStore,
    ScopedStore,
    UserStore,
    type StateStore,
} from "./state.js";
import { MemoryHistoryStore, type HistoryStore } from "./history.js";
import {
    PROTOCOL_VERSION,
    eventToFrames,
    parseIncoming,
    type Frame,
    type IncomingMessage,
    type TextFrame,
} from "./protocol.js";
import { busy, genui, message, runError, type AgentEvent } from "./events.js";
import type {
    ConversationContext,
    Logger,
    PendingInterrupt,
    RunInput,
    Runtime,
    TurnContext,
} from "./runtime.js";
import { runWithTurnContext } from "./emit.js";
import { AuthenticationError, type Authenticator } from "./auth.js";

/** Engine-internal per-conversation record (kept under "conv:{id}:botiva"). */
interface ConversationRecord extends Record<string, unknown> {
    userId: string;
    createdAt: number;
    pendingInterrupt?: PendingInterrupt;
}

interface LiveConnection {
    id: string;
    userId: string;
    conversationId: string;
    meta: Record<string, unknown>;
    deliver(frame: Frame): void | Promise<void>;
}

/** Credential + request material a transport hands the Authenticator. */
export interface AuthInput {
    /** Transport name, e.g. "websocket" | "socket.io". */
    transport?: string;
    /** Credential presented by the client (query token, hello.token, Bearer header). */
    token?: string;
    /** Raw query parameters of the upgrade/handshake request. */
    query?: Record<string, string>;
    /** Raw request headers (lower-cased keys). */
    headers?: Record<string, string>;
}

export interface ConnectParams {
    /** Stable user identity; generated (and returned in `welcome`) when omitted. */
    userId?: string;
    /** Conversation to create or re-attach to; generated when omitted. */
    conversationId?: string;
    /** Highest seq the client already has; history after it is replayed. Default 0 (full replay). */
    watermark?: number;
    /** Transport callback that writes one wire frame to this client. */
    deliver(frame: Frame): void | Promise<void>;
    meta?: Record<string, unknown>;
    /** Credential + request material for the configured Authenticator (if any). */
    auth?: AuthInput;
}

/** Handle a transport holds for one attached client. */
export interface Connection {
    readonly id: string;
    readonly userId: string;
    readonly conversationId: string;
    /** Feed one inbound wire payload (JSON string or parsed frame object). */
    receive(raw: unknown): Promise<void>;
    /** Detach from the conversation (the conversation itself remains resumable). */
    close(): Promise<void>;
}

export interface EngineOptions {
    runtime: Runtime;
    stateStore?: StateStore;
    historyStore?: HistoryStore;
    extensions?: Extension[];
    logger?: Logger;
    /** Sent as a persistent bot message when a conversation is first created. */
    greeting?: string;
    /**
     * Gates connect(): rejects unauthorized attempts (throws AuthenticationError)
     * or replaces the client-asserted userId with a verified one. Omit for the
     * legacy open-door behaviour.
     */
    authenticator?: Authenticator;
}

export class ConversationEngine {
    readonly runtime: Runtime;
    readonly store: StateStore;
    readonly history: HistoryStore;
    readonly extensions: ExtensionRegistry;
    readonly log: Logger;
    #greeting?: string;
    #authenticator?: Authenticator;
    #live = new Map<string, Set<LiveConnection>>();
    #turnLocks = new Set<string>();

    constructor(opts: EngineOptions) {
        if (typeof opts?.runtime?.run !== "function") {
            throw new Error("ConversationEngine requires a runtime implementing run(input, ctx).");
        }
        this.runtime = opts.runtime;
        this.store = opts.stateStore ?? new MemoryStateStore();
        this.history = opts.historyStore ?? new MemoryHistoryStore();
        this.extensions = new ExtensionRegistry(opts.extensions ?? []);
        this.log = opts.logger ?? console;
        this.#greeting = opts.greeting;
        this.#authenticator = opts.authenticator;
    }

    // ── connection lifecycle (called by transport connectors) ───────────────

    async connect(params: ConnectParams): Promise<Connection> {
        if (typeof params?.deliver !== "function") {
            throw new Error("connect() requires a deliver(frame) callback.");
        }

        // Authenticate before touching any state. A configured authenticator can
        // reject the attempt (throws AuthenticationError → transports emit an
        // `error` frame + close) or replace the client-asserted userId with a
        // verified one.
        let userIdHint = params.userId;
        let authClaims: Record<string, unknown> | undefined;
        if (this.#authenticator) {
            const result = await this.#authenticator.authenticate({
                transport: params.auth?.transport ?? "unknown",
                token: params.auth?.token,
                query: params.auth?.query,
                headers: params.auth?.headers,
                userId: params.userId,
                conversationId: params.conversationId,
            });
            if (!result.ok) {
                throw new AuthenticationError(result.reason ?? "unauthorized");
            }
            if (result.userId !== undefined) userIdHint = result.userId;
            authClaims = result.claims;
        }

        const conversationId = params.conversationId ?? this.#id("conv");
        const record = await this.#loadRecord(conversationId, userIdHint);
        const userId = userIdHint ?? record.userId;

        const live: LiveConnection = {
            id: this.#id("connection"),
            userId,
            conversationId,
            meta: authClaims ? { ...(params.meta ?? {}), auth: authClaims } : params.meta ?? {},
            deliver: params.deliver,
        };
        let set = this.#live.get(conversationId);
        if (!set) {
            set = new Set();
            this.#live.set(conversationId, set);
        }
        set.add(live);

        const ctx = this.#conversationContext(conversationId, userId, live.meta);
        if (record.fresh) await this.#notify(() => this.extensions.notifyConversationStart(ctx));
        await this.#notify(() => this.extensions.notifyConnect({ connectionId: live.id }, ctx));

        // 1) welcome (transient) — tells the client its identity + current watermark
        const latest = await this.history.latest(conversationId);
        await live.deliver({
            type: "welcome",
            data: {
                protocol: PROTOCOL_VERSION,
                conversationId,
                userId,
                connectionId: live.id,
                watermark: latest,
            },
        });
        // 2) replay everything the client hasn't seen yet
        const from = params.watermark ?? 0;
        if (latest > from) {
            for (const frame of await this.history.after(conversationId, from)) {
                await live.deliver(frame);
            }
        }
        // 3) greeting — only when the conversation is brand-new (persisted → replays later)
        if (record.fresh && this.#greeting) {
            await this.post(conversationId, message(this.#greeting));
        }

        let closed = false;
        const engine = this;
        return {
            id: live.id,
            userId,
            conversationId,
            async receive(raw: unknown): Promise<void> {
                await engine.#handleInbound(live, raw);
            },
            async close(): Promise<void> {
                if (closed) return;
                closed = true;
                await engine.#disconnect(live);
            },
        };
    }

    async #disconnect(live: LiveConnection): Promise<void> {
        const set = this.#live.get(live.conversationId);
        set?.delete(live);
        const ctx = this.#conversationContext(live.conversationId, live.userId, live.meta);
        await this.#notify(() => this.extensions.notifyDisconnect({ connectionId: live.id }, ctx));
        if (set && set.size === 0) {
            this.#live.delete(live.conversationId);
            await this.#notify(() => this.extensions.notifyConversationEnd(ctx));
        }
    }

    async #handleInbound(live: LiveConnection, raw: unknown): Promise<void> {
        const inbound = parseIncoming(typeof raw === "string" || raw === null ? raw : raw instanceof Uint8Array ? raw.toString() : raw);
        if (!inbound) return;
        if (inbound.kind === "hello") {
            // Handshake happens at connect() time; a late hello is ignored.
            this.log.warn?.("[botiva] late hello frame ignored (handshake happens on connect)");
            return;
        }
        await this.handleMessage(live.conversationId, inbound.message, {
            userId: live.userId,
            origin: live,
        });
    }

    // ── turns ────────────────────────────────────────────────────────────────

    /**
     * Run one turn. Usually invoked via Connection.receive(), but callable
     * directly (e.g. from an HTTP POST endpoint) — events are broadcast to
     * every connection attached to the conversation.
     */
    async handleMessage(
        conversationId: string,
        rawMessage: IncomingMessage,
        opts: { userId?: string; origin?: unknown } = {},
    ): Promise<void> {
        const origin = opts.origin as LiveConnection | undefined;
        const recordStore = this.#recordStore(conversationId);
        const record =
            ((await recordStore.get()) as ConversationRecord | undefined) ??
            (await this.#loadRecord(conversationId, opts.userId)).record;
        const userId = opts.userId ?? record.userId;

        const queue = new AsyncQueue<AgentEvent>();
        const ctx: TurnContext = {
            ...this.#conversationContext(conversationId, userId, origin?.meta ?? {}),
            emit: (event: AgentEvent) => queue.push(event),
        };

        const msg = await this.extensions.applyMessage(rawMessage, ctx);
        if (!msg?.text) {
            queue.close();
            return;
        }

        if (this.#turnLocks.has(conversationId)) {
            queue.close();
            await this.#dispatch(busy(), ctx, { only: origin });
            return;
        }
        this.#turnLocks.add(conversationId);
        try {
            // Echo the user's message to the other attached connections and
            // persist it, so reconnects/other devices get the full transcript.
            const userFrame: TextFrame = {
                type: "text",
                id: msg.id ?? this.#id("msg"),
                from: "user",
                data: { text: msg.text },
                timestamp: Date.now(),
            };
            const userSeq = await this.history.append(conversationId, userFrame);
            this.#broadcast(conversationId, { ...userFrame, seq: userSeq }, origin);

            // Pending interrupt? Then this message is the HITL answer → resume.
            let input: RunInput;
            if (record.pendingInterrupt) {
                input = { resume: msg.text, interrupt: record.pendingInterrupt };
                await recordStore.patch({ pendingInterrupt: undefined });
            } else {
                input = { text: msg.text };
            }

            // Pump runtime events and botivaEmit() events into one stream.
            const pump = (async () => {
                try {
                    await runWithTurnContext(ctx, async () => {
                        for await (const ev of this.runtime.run(input, ctx)) queue.push(ev);
                    });
                } catch (err) {
                    const messageText = err instanceof Error ? err.message : String(err);
                    this.log.error?.(`[botiva] run failed (${conversationId}):`, messageText);
                    queue.push(runError(messageText));
                } finally {
                    queue.close();
                }
            })();

            // GenUI chunks of a turn are grouped under one auto-assigned stream.
            let streamId: string | null = null;
            let streamDone = false;
            let chunkSeq = 0;

            for await (let ev of queue) {
                if (ev.type === "interrupt") {
                    await recordStore.patch({
                        pendingInterrupt: { id: ev.id ?? null, payload: ev.payload ?? null, at: Date.now() },
                    });
                }
                if (ev.type === "genui") {
                    streamId ??= ev.streamId ?? this.#id("stream");
                    ev = {
                        ...ev,
                        streamId: ev.streamId ?? streamId,
                        done: ev.done === true,
                        chunk: { id: ++chunkSeq, ...ev.chunk },
                    };
                    if (ev.done) streamDone = true;
                }
                await this.#dispatch(ev, ctx);
            }
            await pump;

            // Close a genui stream the runtime left open.
            if (streamId && !streamDone) {
                await this.#dispatch(
                    genui(
                        { type: "event", name: "stream_done", payload: null, id: ++chunkSeq },
                        { streamId, done: true },
                    ),
                    ctx,
                );
            }
        } finally {
            this.#turnLocks.delete(conversationId);
        }
    }

    /**
     * Proactive, out-of-turn delivery (reminders, server pushes): runs the
     * event through extensions, persists it and broadcasts it to every
     * connection attached to the conversation.
     */
    async post(conversationId: string, event: AgentEvent, opts: { userId?: string } = {}): Promise<void> {
        const record = (await this.#recordStore(conversationId).get()) as ConversationRecord | undefined;
        const userId = opts.userId ?? record?.userId ?? "system";
        const ctx: TurnContext = {
            ...this.#conversationContext(conversationId, userId, {}),
            emit: (ev: AgentEvent) => void this.post(conversationId, ev, opts),
        };
        await this.#dispatch(event, ctx);
    }

    // ── internals ────────────────────────────────────────────────────────────

    async #dispatch(
        ev: AgentEvent,
        ctx: TurnContext,
        opts: { only?: LiveConnection } = {},
    ): Promise<void> {
        const out = await this.extensions.applyEvent(ev, ctx);
        if (!out) return;
        for (const { frame, persistent } of eventToFrames(out, (prefix) => this.#id(prefix))) {
            let f = frame;
            if (persistent) {
                const seq = await this.history.append(ctx.conversationId, f);
                f = { ...f, seq } as Frame;
            }
            if (opts.only) {
                try {
                    await opts.only.deliver(f);
                } catch (err) {
                    this.log.warn?.("[botiva] deliver failed:", err);
                }
            } else {
                this.#broadcast(ctx.conversationId, f);
            }
        }
    }

    #broadcast(conversationId: string, frame: Frame, except?: LiveConnection): void {
        const set = this.#live.get(conversationId);
        if (!set) return;
        for (const conn of set) {
            if (conn === except) continue;
            try {
                void conn.deliver(frame);
            } catch (err) {
                this.log.warn?.("[botiva] deliver failed:", err);
            }
        }
    }

    async #loadRecord(
        conversationId: string,
        preferredUserId?: string,
    ): Promise<{ record: ConversationRecord; fresh: boolean; userId: string }> {
        const recordStore = this.#recordStore(conversationId);
        let record = (await recordStore.get()) as ConversationRecord | undefined;
        const fresh = !record;
        if (!record) {
            record = { userId: preferredUserId ?? this.#id("user"), createdAt: Date.now() };
            await recordStore.set(record);
        }
        return { record, fresh, userId: record.userId };
    }

    #recordStore(conversationId: string): ScopedStore<ConversationRecord> {
        return new ScopedStore<ConversationRecord>(this.store, `conv:${conversationId}:botiva`);
    }

    #conversationContext(
        conversationId: string,
        userId: string,
        meta: Record<string, unknown>,
    ): ConversationContext {
        return {
            conversationId,
            userId,
            userStore: new UserStore(this.store, userId),
            conversationStore: new ConversationStore(this.store, conversationId),
            log: this.log,
            meta,
        };
    }

    async #notify(fn: () => Promise<void>): Promise<void> {
        try {
            await fn();
        } catch (err) {
            this.log.warn?.("[botiva] extension hook failed:", err);
        }
    }

    #id(prefix: string): string {
        return `${prefix}-${randomUUID()}`;
    }
}

File: packages/core/python/botiva/engine.py
"""ConversationEngine — Python port of the botiva engine.

Same responsibilities as @botiva/core: identity (user/conversation/connection),
watermark replay, per-conversation turn lock, HITL pending interrupts,
multi-connection fan-out, extension chain, GenUI stream grouping.

The ambient turn context uses ``contextvars`` — the asyncio counterpart of
Node's AsyncLocalStorage (PROTOCOL.md §9):

    from botiva import botiva_emit, botiva_context, ui

    async def my_langgraph_node(state):        # no plumbing required
        botiva_emit(ui("weather-card", {"temp": 22}))
        ctx = botiva_context()
        await ctx.user_store.patch({"seen": True})
"""

from __future__ import annotations

import contextvars
import logging
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Awaitable, Callable, Protocol

from .events import AgentEvent, busy as busy_event, genui, message, run_error
from .protocol import (
    PROTOCOL_VERSION,
    Frame,
    IncomingMessage,
    event_to_frames,
    parse_incoming,
)
from .state import (
    ConversationStore,
    HistoryStore,
    MemoryHistoryStore,
    MemoryStateStore,
    ScopedStore,
    StateStore,
    UserStore,
)

logger = logging.getLogger("botiva")


def _new_id(prefix: str) -> str:
    return f"{prefix}-{uuid.uuid4()}"


def _now_ms() -> int:
    return int(time.time() * 1000)


@dataclass
class TurnContext:
    conversation_id: str
    user_id: str
    user_store: UserStore
    conversation_store: ConversationStore
    meta: dict[str, Any] = field(default_factory=dict)
    _emit: Callable[[AgentEvent], Awaitable[None]] | None = None

    async def emit(self, event: AgentEvent) -> None:
        """Push an out-of-band event into the current turn."""
        if self._emit is not None:
            await self._emit(event)


class Runtime(Protocol):
    """The driving port — the only thing an agent framework adapter implements."""

    def run(self, input: dict[str, Any], ctx: TurnContext) -> AsyncIterator[AgentEvent]: ...


class Extension(Protocol):
    """Optional hooks; return None from on_message/on_event to swallow/drop."""

    name: str
    # async def on_message(self, msg, ctx) -> IncomingMessage | None
    # async def on_event(self, ev, ctx) -> AgentEvent | None
    # async def on_conversation_start/end(self, ctx) / on_connect/on_disconnect(...)


# ── ambient turn context (contextvars) ───────────────────────────────────────

_current_turn: contextvars.ContextVar[TurnContext | None] = contextvars.ContextVar(
    "botiva_turn", default=None
)


def botiva_context() -> TurnContext | None:
    """The TurnContext of the currently executing turn, if any."""
    return _current_turn.get()


def botiva_emit(event: AgentEvent) -> bool:
    """Emit into the current turn (fire-and-forget); False outside a turn.

    Events are queued synchronously and dispatched by the engine's turn loop.
    """
    ctx = _current_turn.get()
    if ctx is None or ctx._emit is None:
        return False
    pending = getattr(ctx, "_pending_emits", None)
    if pending is None:
        return False
    pending.append(event)
    return True


@dataclass
class Connection:
    """Handle a transport holds for one attached client."""

    id: str
    user_id: str
    conversation_id: str
    _engine: "ConversationEngine"
    _live: "_LiveConnection"
    _closed: bool = False

    async def receive(self, raw: Any) -> None:
        """Feed one inbound wire payload (JSON string, bytes or frame dict)."""
        inbound = parse_incoming(raw)
        if inbound is None:
            return
        if inbound.hello is not None:
            logger.warning("[botiva] late hello frame ignored (handshake happens on connect)")
            return
        await self._engine.handle_message(
            self.conversation_id, inbound.message, user_id=self.user_id, origin=self._live
        )

    async def close(self) -> None:
        """Detach; the conversation itself stays resumable."""
        if self._closed:
            return
        self._closed = True
        await self._engine._disconnect(self._live)


@dataclass(eq=False)  # identity semantics: lives in sets, compared with `is`
class _LiveConnection:
    id: str
    user_id: str
    conversation_id: str
    meta: dict[str, Any]
    deliver: Callable[[Frame], Any]


class ConversationEngine:
    def __init__(
        self,
        runtime: Runtime,
        *,
        state_store: StateStore | None = None,
        history_store: HistoryStore | None = None,
        extensions: list[Any] | None = None,
        greeting: str | None = None,
    ) -> None:
        self.runtime = runtime
        self.store: StateStore = state_store or MemoryStateStore()
        self.history: HistoryStore = history_store or MemoryHistoryStore()
        self.extensions = extensions or []
        self.greeting = greeting
        self._live: dict[str, set[_LiveConnection]] = {}
        self._turn_locks: set[str] = set()

    # ── connection lifecycle ────────────────────────────────────────────────

    async def connect(
        self,
        deliver: Callable[[Frame], Any],
        *,
        user_id: str | None = None,
        conversation_id: str | None = None,
        watermark: int = 0,
        meta: dict[str, Any] | None = None,
    ) -> Connection:
        conversation_id = conversation_id or _new_id("conv")
        record, fresh = await self._load_record(conversation_id, user_id)
        user_id = user_id or record["userId"]

        live = _LiveConnection(_new_id("connection"), user_id, conversation_id, meta or {}, deliver)
        self._live.setdefault(conversation_id, set()).add(live)

        ctx = self._context(conversation_id, user_id, live.meta)
        if fresh:
            await self._notify("on_conversation_start", ctx)
        await self._notify("on_connect", live.id, ctx)

        # 1) welcome (transient)
        latest = await self.history.latest(conversation_id)
        await self._deliver(live, {
            "type": "welcome",
            "data": {
                "protocol": PROTOCOL_VERSION,
                "conversationId": conversation_id,
                "userId": user_id,
                "connectionId": live.id,
                "watermark": latest,
            },
        })
        # 2) replay everything the client hasn't seen
        if latest > watermark:
            for frame in await self.history.after(conversation_id, watermark):
                await self._deliver(live, frame)
        # 3) greeting on brand-new conversations
        if fresh and self.greeting:
            await self.post(conversation_id, message(self.greeting))

        return Connection(live.id, user_id, conversation_id, self, live)

    async def _disconnect(self, live: _LiveConnection) -> None:
        conns = self._live.get(live.conversation_id)
        if conns:
            conns.discard(live)
        ctx = self._context(live.conversation_id, live.user_id, live.meta)
        await self._notify("on_disconnect", live.id, ctx)
        if conns is not None and not conns:
            self._live.pop(live.conversation_id, None)
            await self._notify("on_conversation_end", ctx)

    # ── turns ───────────────────────────────────────────────────────────────

    async def handle_message(
        self,
        conversation_id: str,
        raw_message: IncomingMessage,
        *,
        user_id: str | None = None,
        origin: Any = None,
    ) -> None:
        record_store = self._record_store(conversation_id)
        record, _ = await self._load_record(conversation_id, user_id)
        user_id = user_id or record["userId"]

        ctx = self._context(conversation_id, user_id, getattr(origin, "meta", {}) or {})

        # Turn-local GenUI grouping + pending emit queue for botiva_emit().
        turn = {"stream_id": None, "stream_done": False, "chunk_seq": 0}
        pending_emits: list[AgentEvent] = []
        ctx._pending_emits = pending_emits  # type: ignore[attr-defined]

        async def dispatch(ev: AgentEvent, only: _LiveConnection | None = None) -> None:
            if ev.get("type") == "interrupt":
                await record_store.patch({"pendingInterrupt": {
                    "id": ev.get("id"), "payload": ev.get("payload"), "at": _now_ms(),
                }})
            if ev.get("type") == "genui":
                turn["stream_id"] = turn["stream_id"] or ev.get("streamId") or _new_id("stream")
                turn["chunk_seq"] += 1
                chunk = dict(ev.get("chunk") or {})
                chunk.setdefault("id", turn["chunk_seq"])
                ev = {**ev, "streamId": ev.get("streamId") or turn["stream_id"],
                      "done": ev.get("done") is True, "chunk": chunk}
                if ev["done"]:
                    turn["stream_done"] = True
            out: AgentEvent | None = ev
            for ext in self.extensions:
                handler = getattr(ext, "on_event", None)
                if handler is None or out is None:
                    continue
                out = await handler(out, ctx)
                if out is None:
                    return
            for mapping in event_to_frames(out, _new_id):
                frame = mapping.frame
                if mapping.persistent:
                    seq = await self.history.append(conversation_id, frame)
                    frame = {**frame, "seq": seq}
                if only is not None:
                    await self._deliver(only, frame)
                else:
                    await self._broadcast(conversation_id, frame)

        async def drain_emits() -> None:
            while pending_emits:
                await dispatch(pending_emits.pop(0))

        ctx._emit = dispatch  # explicit ctx.emit() path

        msg: IncomingMessage | None = raw_message
        for ext in self.extensions:
            handler = getattr(ext, "on_message", None)
            if handler is None or msg is None:
                continue
            msg = await handler(msg, ctx)
            if msg is None:
                return
        if not msg.text:
            return

        if conversation_id in self._turn_locks:
            await dispatch(busy_event(), only=origin)
            return
        self._turn_locks.add(conversation_id)
        token = _current_turn.set(ctx)
        try:
            # Persist + fan out the user's message (everyone but the sender).
            user_frame: Frame = {
                "type": "text", "id": msg.id or _new_id("msg"), "from": "user",
                "data": {"text": msg.text}, "timestamp": _now_ms(),
            }
            seq = await self.history.append(conversation_id, user_frame)
            await self._broadcast(conversation_id, {**user_frame, "seq": seq}, except_=origin)

            # Pending interrupt? Then this message is the HITL answer.
            if record.get("pendingInterrupt"):
                run_input = {"resume": msg.text, "interrupt": record["pendingInterrupt"]}
                await record_store.patch({"pendingInterrupt": None})
            else:
                run_input = {"text": msg.text}

            try:
                async for ev in self.runtime.run(run_input, ctx):
                    await drain_emits()
                    await dispatch(ev)
                await drain_emits()
            except Exception as err:  # noqa: BLE001 — surfaced to the client
                logger.exception("[botiva] run failed (%s)", conversation_id)
                await dispatch(run_error(err))

            # Close a genui stream the runtime left open.
            if turn["stream_id"] and not turn["stream_done"]:
                turn["chunk_seq"] += 1
                await dispatch(genui(
                    {"type": "event", "name": "stream_done", "payload": None,
                     "id": turn["chunk_seq"]},
                    stream_id=turn["stream_id"], done=True,
                ))
        finally:
            _current_turn.reset(token)
            self._turn_locks.discard(conversation_id)

    async def post(self, conversation_id: str, event: AgentEvent, *, user_id: str | None = None) -> None:
        """Proactive, out-of-turn delivery (reminders, server pushes)."""
        record = await self._record_store(conversation_id).get() or {}
        user_id = user_id or record.get("userId") or "system"
        ctx = self._context(conversation_id, user_id, {})
        out: AgentEvent | None = event
        for ext in self.extensions:
            handler = getattr(ext, "on_event", None)
            if handler is None or out is None:
                continue
            out = await handler(out, ctx)
            if out is None:
                return
        for mapping in event_to_frames(out, _new_id):
            frame = mapping.frame
            if mapping.persistent:
                seq = await self.history.append(conversation_id, frame)
                frame = {**frame, "seq": seq}
            await self._broadcast(conversation_id, frame)

    # ── internals ───────────────────────────────────────────────────────────

    async def _broadcast(self, conversation_id: str, frame: Frame, except_: Any = None) -> None:
        for conn in list(self._live.get(conversation_id, ())):
            if conn is except_:
                continue
            await self._deliver(conn, frame)

    async def _deliver(self, live: _LiveConnection, frame: Frame) -> None:
        try:
            result = live.deliver(frame)
            if hasattr(result, "__await__"):
                await result
        except Exception:  # noqa: BLE001
            logger.warning("[botiva] deliver failed", exc_info=True)

    async def _load_record(
        self, conversation_id: str, preferred_user_id: str | None
    ) -> tuple[dict[str, Any], bool]:
        record_store = self._record_store(conversation_id)
        record = await record_store.get()
        if record is not None:
            return dict(record), False
        record = {"userId": preferred_user_id or _new_id("user"), "createdAt": _now_ms()}
        await record_store.set(record)
        return record, True

    def _record_store(self, conversation_id: str) -> ScopedStore:
        return ScopedStore(self.store, f"conv:{conversation_id}:botiva")

    def _context(self, conversation_id: str, user_id: str, meta: dict[str, Any]) -> TurnContext:
        return TurnContext(
            conversation_id=conversation_id,
            user_id=user_id,
            user_store=UserStore(self.store, user_id),
            conversation_store=ConversationStore(self.store, conversation_id),
            meta=meta,
        )

    async def _notify(self, hook: str, *args: Any) -> None:
        for ext in self.extensions:
            handler = getattr(ext, hook, None)
            if handler is None:
                continue
            try:
                result = handler(*args)
                if hasattr(result, "__await__"):
                    await result
            except Exception:  # noqa: BLE001
                logger.warning("[botiva] extension hook %s failed", hook, exc_info=True)
